[iOS][#90] 카드 수정 기능을 구현하고 JK의 피드백을 반영하였습니다.

ehgud0670 opened the pull request:
> issue: #90 인 카드 수정 기능을 구현하였습니다. 
  * 기존 코드는 새로운 카드를 추가하는 역할을 하는 `CardViewController`만 있었지만, 수정 기능을 구현하기 위해 `CardViewController`는 카드의 제목, 내용을 입력하는 **부모 뷰컨트롤러의 역할**을 하게 하고, 
이 뷰컨트롤러를 **상속**하는 `NewCardViewController`, `EditingCardViewController`를 만들어 `NewCardViewController`는 **새로운 카드를 만드는 역할**, `EditingCardViewController`는 **기존의 카드를 수정하는(실제로는 새 카드를 만드는) 역할**을 하도록 하였습니다.
    * 이를 위해 CardViewController의 프로퍼티인 `CreateButton`, `TitleField`, `ContentView` 객체의 접근 제한자를 `fileprivate`으로 두었습니다. 
    * 프로토콜 CardCreatable을 만들어 `CardViewController`를 상속한 객체만 채택할 수 있도록 하였습니다.
    * `CardViewController`에서 `CreateButton` 객체에 addTarget을 설정하는 기존 코드를 지우고 addTarget을 `NewCardViewController`와 `EditingCardViewController`에서 설정하도록 하였습니다.

> JK의 피드백을 반영하였습니다. (pull request: https://github.com/codesquad-member-2020/todo-10/pull/73)

* 다음 프로젝트에도 재 사용가능한 NetworkManager가 되도록 하시라는 피드백을 받고, NetworkManager 객체 안에  enum으로 두었던 EndPoints를 다른 파일에 두어 따로 분리하였습니다.
* 들여쓰기가 깊다는 코드 부분을 들여쓰기 한단계만 없어지도록 하였습니다. 메소드의 역할이 두 개라서 분리하였습니다. ( MainViewController 객체의 configureColumnsCase 메소드 입니다 )
* ColumnDataSource객체가 컬럼 뷰모델 역할을 할 수 있다는 JK의 피드백을 듣고 ColumnDataSource객체가 CardViewModel 배열을 갖는 기존의 코드를 고수하였습니다. 사실 그래서 ColumnDataSource가 뷰모델 프로토콜을 채택하게 하고 싶었는데 막상 컬럼 뷰모델로서 뷰에 입히는 데이터 값이 없어서 하지 않았습니다(수업 시간에 말씀해주신 것처럼 ). 대신 서버와 통신할 때 columnID가 필요해서 속성으로 갖게 했습니다. (하지만 지금 생각하니 columnID 속성이 뷰에 대입할 데이터가 아니라면 뷰컨트롤러가 가지고 있는것이 나을 것같습니다.)
*  또 ColumnDataSource 객체가 cardID를 반환하는 메소드를 없애고 cardViewModel를 반환하는 메소드를 만들어 CardVIewModel 객체 자체를 보내라는 피드백을 반영하였습니다.

ehgud0670 committed:
[#90] refactor: 앞으로 편집 기능을 구현하기 위해 접근제한자를 fileprivate으로 둠

ehgud0670 committed:
[#90] refactor: 삭제하는 행동에 대한 delay를 contextMenu인 경우에만 주기 위해 코드 수정함
		이유: 메뉴 dismiss 하는 애니메이션과 겹치지 않기 위해 delay를 주었던 건데
		      그 결과 swipe하여 행을 delete하는 애니메이션 너무 느려져서 delay를
		      매개변수로 추가함으로써 contextMenu에서 행을 삭제하는 경우에만 delay를 주도록 하였다.

ehgud0670 committed:
[#90] style: Delegate 관련된 프로토콜이므 Delegate 관련 프로토콜임을
	프로토콜 명을 수정하여 명시하였고, 안의 델리게이트 메서드는 네임스페이스로 구분함

ehgud0670 committed:
[#90] style: Delegate 메서드를 프로토콜명으로 네임스페이스를 구분하도록 메소드명 수정함

ehgud0670 committed:
[#90] refactor: 상속과 제한을 설정한 프로토콜, fileprivate을 이용해 카드 생성을 하도록 리팩토링함
		이렇게 리팩토링한 이유: 이후의 기존 카드를 편집하는 기능을 깔끔하게 구현하기 위해서

ehgud0670 committed:
[#90] feat: 카드를 편집하는 뷰 컨트롤러를 생성함
	    또 카드 편집하는 것을 서버에 request하기 위해 EditingUseCase를 만들었고,
	    일단 테스트하기 위해 mock 객체와 stub을 만들었다.
	=> 근데 CreateViewController를 복붙해서 뼈대만 만든거지 수정하거나 추가해야 될 부분이 있다.

ehgud0670 committed:
[#90] style: 콜론 왼쪽에 공백 없앰 (스타일을 지키기 위해서)

ehgud0670 committed:
[#90] refactor: cardID 메소드 삭제하고 뷰모델을 반환하는 cardViewModel를
		사용하도록 리팩토링함
	이유: DataSource가 cardID를 반환하는 메소드를 갖는건 책임에 맞지않다.
	      따라서 어차피 뷰모델이 필요하니 뷰모델을 리턴하는 메소드로 대체하였다.

ehgud0670 committed:
[#90] style: 해당 메소드는 응답을 받아서 성공했는지 결과를 확인하는게 목적이라서
	     makeDeleteResult 라고 수정함

	     이유:해당 메소드명은 응답은 만드는게 아니라서 틀렸다.
		  그리고 응답을 받는 것만이 목적이 아니고 응답 안에 status를 끄집어내서 성공("SUCCESS")이면 서버에서
                  해당 데이터가 성공적으로 삭제되었음을 확인하는 게 목적이다.
		  따라서 함수명을 makeDeleteResult가 적당해 수정하였다.

ehgud0670 committed:
[#90] refactor: 메소드의 역할이 카드 생성을 request하고 응답을 받았을때
		뷰모델 객체를 만드는 것이라 정하고 클래스명, 함수명, 리턴타입을 수정함

ehgud0670 committed:
[#90] refactor: 카드를 수정하는 request이므로 http방식을 patch로 수정하고
		뷰모델을 만드는 목적에 맞게 클래스 이름, 메소드명을 수정함

ehgud0670 committed:
[#90] refactor: delegate 객체에서 데이터소스를 참조하는 것은 옳지 않으므로
      swipe로 delete 이벤트를 받았을때 모델을 데이터 소스를 가지고 있는 뷰 컨트롤러에서
	바꾸도록 노티를 보내는 방식으로 리팩토링함

ehgud0670 committed:
[#90] refactor: 1. menu로 delete하는 이벤트를 받으면 노티하고 해당 모델을
		   컨트롤러에서 삭제하도록 리팩토링함
      		2. delegate는 여러개의 tableView와 관계를 가질 수 있는 1:n관계이므로,
		   지금은 뷰컨트롤러에 테이블뷰가 하나밖에 없어서 상관없지만 뷰컨에 또 다른 테이블뷰가 생길 것을 대비해
		   테이블 뷰를 userInfo에 함께 보내서 보낸 tableView의 행을 삭제하도록 함
		3. 이전 addOberver에서 #selector 함수는 notification을 핸들하기 어려우므로 notification을 매개변수로 가지는
		   다른 addObserver 를 사용함

ehgud0670 committed:
[#90] feat: 카드 수정되도록 기능 구현함. 서버 연동도 잘됨

ehgud0670 committed:
[#90] refactor: 수정하는 경우는 이미 content가 있는 경우이므로
		커서의 위치를 처음으로 두지 않도록 수정하고,
		텍스트의 색을 검게하도록 기능 수정함

ehgud0670 committed:
[#90] refactor: 카드 화면 생성시 placeHolder 설정하는 것을 새로운 카드 만드는 경우에만
		하도록 리팩토링함

ehgud0670 committed:
[#90] style: 네트워크의 응답을 찾는 객체의 이름을 더 목적에 맞게 변경함

ehgud0670 committed:
[#90] refactor: columnViewController를 만드는 함수를 만들어서 분리함
		이유: 메소드는 하나의 역할만 해야 한다.

ehgud0670 committed:
[#90] refactor: Foundation 은 UIKit에서 포함하고 있으므로 import 문 삭제함

ehgud0670 committed:
[#90] refactor: 다음 프로젝트에도 NetworkManager를 재사용 가능하도록 EndPoints를 분리함

ehgud0670 committed:
[#90] 뷰모델 객체를 찾기 쉽도록 파일을 나눔

ehgud0670 committed:
[#90] refactor: ColumnViewController가 Column모델 객체를 따로 갖지 않고,
ColumnDataSource가 뷰모델로서 columnID와 CardViewModel 배열을 갖도록 수정 및 적용함

ColumnDataSource는 뷰모델로서 모델의 값을 업데이트하고 이를 뷰에 반영하는 역할을 한다.

ehgud0670 committed:
[#90] refactor: 상태들을 캡슐화하고 상태들에게 바꿀 값만 행동으로 바꾸도록 메소드 구현함

ehgud0670 committed:
[#90] chore:  NewCard 객체는 모델객체이므로 객체를 파일을 만들어 Models에 둠

ehgud0670 committed:
[#90] chore: 클래스가 enum, 프로퍼티, 생성자, 연산프로퍼티, 메서드 순으로
	갖도록 조정함

godrm commented on iOS/Todo/Todo/Coder/ColumnsUseCase.swift:
```diff
@@ -19,7 +19,7 @@ enum LoginInfo {
 struct ColumnsUseCase {
     static func makeColumns(with manager: NetworkManagable, completed: @escaping (ColumnsDataSource?) -> ()) {
-        try? manager.requestResource(from: NetworkManager.EndPoints.columns, method: .post,
+        try? manager.requestResource(from: EndPoints.columns, method: .post,
```
유스케이스가 고정된다면 엔드포인트도 결국 유스케이스와 관련이 있나 고민해보세요.

ehgud0670 replied:
넵 ColumnsUseCase에 고정되는 엔드 포인트 `columns`를 고정해놓았습니다. 감사합니다.

godrm commented on iOS/Todo/Todo/Coder/DeleteUseCase.swift:
```diff
@@ -9,16 +9,16 @@
+    static func makeDeleteResult(columnID: Int,
                                    cardID: Int,
                                    with manager: NetworkManagable,
                                    result: @escaping (Bool?) -> ()) {
```
Response나 Result를 의식할 필요는 없어보여요. 그냥 Delete를 요청하는거죠.

ehgud0670 replied:
음 그러면 다른 유스케이스 객체는 응답으로 무언가를 만들어서 메소드 이름이 make로 시작하지만 이 경우에는 `requestDelete`라고 메소드 명을 짓는 것이 적절할까요?

godrm replied:
무엇이 목적이냐는거죠. 응답 객체를 생성하려는 게 아니잖아요. 반드시 requestDelete로 하라는게 아니라 makeDeleteResult 보다는 낫네요.

godrm commented on iOS/Todo/Todo/Coder/DeleteUseCase.swift:
```diff
@@ -9,16 +9,16 @@
                                    with manager: NetworkManagable,
                                    result: @escaping (Bool?) -> ()) {
-        try? manager.requestResource(from: "\(NetworkManager.EndPoints.column)/\(columnID)/card/\(cardID)",
+        try? manager.requestResource(from: "\(EndPoints.column)/\(columnID)/card/\(cardID)",
```
이정도면 엔드 포인트도 조합하거나 생성해는 주는 객체가 필요할 것 같네요.

ehgud0670 replied:
넵 `EndPointFactory` 객체를 만들어서 이미 존재하는 카드와 관련된 앤드포인트를 만드는 메소드와 
새로 만들 카드와 관련 앤드포인트를 만드는 메소드를 구현하여 앤드포인트를 생성하도록 하였습니다.

godrm commented on iOS/Todo/Todo/Coder/DeleteUseCase.swift:
```diff
@@ -9,16 +9,16 @@
-                guard let cardResponse = try? JSONDecoder().decode(CardResponse.self, from: data) else { return }
+                guard let cardResponse = try? JSONDecoder().decode(DeleteResponse.self, from: data) else { return }
                 guard cardResponse.status == .success else { return }
                 result(true)
```
유즈케이스가 단지 디코딩까지만 하는게 아니라 좀 더 적극적으로 연동할 수 있는지 고민해보세요.

ehgud0670 replied:
넵 근데 여기서는 디코드의 결과만 알면 되서 더 연동할 수 있을지 잘 모르겠습니다. ㅠ
제가 만든 다른 유즈케이스는 응답으로 뷰모델 객체를 만드는 작업을 할 수 있는데 
DeleteUseCase의 메소드의 경우에는 성공("SUCCESS") 적으로 삭제가 되었음을 확인만 하면 되는 거라 
잘 모르겠습니다. ( 그래서 삭제의 경우에는 유즈케이스 객체인건지 잘 모르겠네요 ㅜ 유즈케이스도 잘 몰라서 플젝 이후에 그 의미를 자세히 알아보려고 합니다!)

* 삭제 요청에 대한  서버 응답입니다
> 성공 
```swift
{
    "status": "SUCCESS",
    "content": "Card is Successfully Deleted"
}
```
> 실패
```swift
{
    "status": "ERROR",
    "content": "Resource Not Found"
}
```

godrm replied:
모르겠으면 만들지 말고 그냥 뷰 컨트롤러에 구현하세요.

ehgud0670 replied:
넵

godrm commented on iOS/Todo/Todo/Delegates/ColumnTableDelegate.swift:
```diff
@@ -58,15 +45,27 @@ extension ColumnTableDelegate {
+        NotificationCenter.default.post(name: Notification.menuDeleteEventOccured,
+                                        object: self,
+                                        userInfo: ["tableView": tableView, "indexPath" : indexPath])
+    }
```
지금처럼 ColumnTableDelegate에서 이벤트에 대한 노티만 보낼꺼라면, ViewController가 UITableViewDelegate 프로토콜을 채택해도 됩니다. 특히 옵저버가 하나인 경우에는 그렇게 해도 됩니다.

ehgud0670 replied:
넵 특히 옵저버가 하나인 경우라서 ViewController가 UITableViewDelegate를 채택하는 구조로 반영하였습니다. 이 구조가 훨씬 단순하고 유연하여 좋은 것 같습니다. 감사합니다.

godrm commented on iOS/Todo/Todo/Models/ColumnTableDataSource.swift:
```diff
@@ -29,19 +24,27 @@ final class ColumnTableDataSource: NSObject {
+    func update(cardViewModel: CardViewModel, at index: Int) {
+        guard index < cardViewModels.count else { return }
+        cardViewModels[index] = cardViewModel
     }
```
이렇게 윗부분만 보니까 ColumnTableDataSource 객체는 데이터 소스라기 보다는 ColumnViewModel 인 것 같네요. 
- 이 객체는 데이터 소스 역할을 위해서 만든걸까요? 
- 뷰 모델을 관리하는 객체인데 데이터 소스 역할을 하는걸까요?

ehgud0670 replied:
아 지금 저는 뷰모델을 관리하는 객체인데 데이터 소스 역할도 하는걸로 생각하여 구현하였습니다.

ehgud0670 replied:
UITableVIewDataSource 객체는 테이블 뷰 관련 데이터를 관리하고 셀에 데이터를 넣는 역할이라 
모델 데이터를 뷰에 바인딩하는 역할을 하는 뷰모델이 데이터소스의 역할도 할 수 있지 않을까 싶어 이런식으로 구현했습니다.

godrm commented on iOS/Todo/Todo/Models/ColumnTableDataSource.swift:
```diff
@@ -29,19 +24,27 @@ final class ColumnTableDataSource: NSObject {
+        cardViewModels[index] = cardViewModel
     }
 }
 
```
아래 `tableView(_ tableView:, cellForRowAt:)` 에 fatalError도 지양하세요. 셀 안 만들어진다고 앱이 죽으면 안 됩니다.

ehgud0670 replied:
넵 명심하겠습니다. 앱은 죽지 않게 하고 버그를 고치는 것은 그 이후에 해야 한다는 것을 유의하며 코드를 작성하도록 하겠습니다. 코드 반영하였습니다!

godrm commented on iOS/Todo/Todo/ViewControllers/CardViewController.swift:
```diff
@@ -106,3 +93,61 @@ final class CardViewController: UIViewController {
         contentView.heightAnchor.constraint(equalTo: view.heightAnchor, multiplier: 0.5).isActive = true
     }
 }
```
부모 컨트롤러도 별도 파일로 분리하세요.

ehgud0670 replied:
넵 별도의 파일로 분리하였고 프로퍼티들도 모두 private으로 두어 메소드로 메시지를 주고 받도록 구현하였습니다. 

근데 이렇게 피드백 하신 의중이 궁금합니다. 제가 생각한 바로는 
1. 파일로 분리하지 않으면 상속할때 마다 파일이 너무 커져버려서 인가요?
2. fileprivate은 file안에서 캡슐화되지 않기 때문에 지양하고 private을 사용하라고 파일을 나누라고 하신 말씀이신가요?

godrm commented on iOS/Todo/Todo/ViewControllers/CardViewController.swift:
```diff
@@ -106,3 +93,61 @@ final class CardViewController: UIViewController {
 }
+
+@objc protocol CardCreatable where Self: CardViewController {
+    func createCard()
```
상속 관계를 만들꺼면 createCard()가 super-class에 선언되면 되지 않나요?

ehgud0670 replied:
넵 맞습니다. 하지만 부모 클래스에서는 creatCard()가 비어있게 되어 프로토콜을 사용하였습니다. 
반영하여 부모 클래스의 creatCard()는 비어있는데 괜찮을까요??

godrm commented on iOS/Todo/Todo/ViewControllers/CardViewController.swift:
```diff
@@ -106,3 +93,61 @@ final class CardViewController: UIViewController {
+        guard let cardData = try? JSONEncoder().encode(NewCard(title: titleField.text, content: content)) else { return }
+        NewCardViewModelUseCase.makeNewCardViewModel(columnID: columnID,
+                                                     cardData: cardData, with: MockCardCreateSuccessStub()) { cardViewModel in
+                                                        guard let cardViewModel = cardViewModel else { return }
```
유즈 케이스를 사용하는 이유는 뷰 컨트롤러가 Mock인지 Stub인지 구분없이 호출하기 위해서 입니다.
지금처럼 작성하면 정상 서버로 바꾸면 뷰 컨트롤러도 바꿔야 겠네요.

ehgud0670 replied:
제가 피드백을 이해못했습니다ㅠ 다시 한번 말씀해주실수 있을까요?
1. 일단 저의 코드는 실제 서버에 데이터를 로드하려면 모든 유즈케이스에 일일히 NetworkManager()를 집어넣어야 하고, 
Mock데이터를 받으려면 또 모든 유즈케이스에 일일히 모크스텁객체를 집어넣어야 하는 구조입니다.

2. 또 한 컬럼의 화면에서 추가(+) 버튼을 누를때마다 NewCardViewController가 생성되는 구조이고, 편집 버튼을 누를때마다 해당 카드에 대한 EditingCardViewController가 생성되는 구조라서 뷰컨트롤러를 바꿔주는 건 없습니다..(음 제가 생각하기로는요)

godrm replied:
말한것처럼 유즈 케이스에서만 바꾸면 되는데 뷰 컨트롤러에서도 바꿔야 하는 상수나 코드가 있다면 분리가 잘 못 된거라는 겁니다.

godrm commented on iOS/Todo/Todo/ViewControllers/CardViewController.swift:
```diff
@@ -106,3 +93,61 @@ final class CardViewController: UIViewController {
+    @objc func createCard() {
+        guard let columnID = columnID else { return }
+        guard let content = contentView.text else { return }
+        guard let cardData = try? JSONEncoder().encode(NewCard(title: titleField.text, content: content)) else { return }
```
인코딩도 여기서 할 필요가 없습니다. 데이터 흐름만 제어하고 구체적인 동작은 하위 객체에서 해야겠죠.

ehgud0670 replied:
넵 반영하였습니다. 피드백 감사합니다.

godrm commented on iOS/Todo/Todo/ViewControllers/CardViewController.swift:
```diff
@@ -106,3 +93,61 @@ final class CardViewController: UIViewController {
+                DispatchQueue.main.async {
+                    self.titleField.text = card?.title
+                    self.contentView.text = card?.content
+                }
```
여기는 뭘까요? 뷰모델에 뷰 모델이 있는건가요? 뷰 모델이 바뀌어서 바인드를 매번 바꿔야 하는건가요?

ehgud0670 replied:
음 사실 뷰모델이 아닌데 제가 잘못 이름을 지었습니다. 단지 테이블의 row을 가지고 있으면 이후에 뷰모델 배열을 업데이트 할때 index로 바로 찾아서 업데이트를 할 수 있어서 row값을 집어넣으려고 또 다른 객체를 만든 것입니다. 
피드백을 받고 EditingCardViewController에 row 값과 뷰 모델 객체를 프로퍼티로 갖는 구조로 하였습니다. 
(EditingCardViewController는 특정한 행의 카드를 수정하는 컨트롤러이니 row를 가져도 된다고 판단하여 프로퍼티로 두었습니다.)

godrm commented on iOS/Todo/Todo/ViewControllers/ColumnViewController.swift:
```diff
@@ -52,56 +53,100 @@ final class ColumnViewController: UIViewController {
+        present(editingCardViewController, animated: true)
+    }
+    
+    func cardViewControllerDidCardCreate(_ cardViewModel: CardViewModel) {
+        columnTableDataSource.append(cardViewModel: cardViewModel)
+        DispatchQueue.main.async {
+            self.columnTable.reloadData()
+        }
+    }
+    
+    func cardViewControllerDidCardEdit(_ willEditCardViewModel: WillEditCardViewModel) {
+        columnTableDataSource.update(cardViewModel: willEditCardViewModel.cardViewModel,
+                                     at: willEditCardViewModel.row)
         DispatchQueue.main.async {
             self.columnTable.reloadData()
         }
```
여기 두 메소드 cardViewControllerDidCardCreate, cardViewControllerDidCardEdit 에서는 모델이나 값을 바꾸는 걸까요? 
화면을 같이 갱신해야만 하는걸까요?

ehgud0670 replied:
아닙니다. 모델이나 값을 바꾸는 메소드(정확히는 뷰모델 객체에게 바꿔달라고 메시지를 보내는 역할입니다)인거고 화면을 갱신하면 안됩니다. 
화면을 갱신하는 것은 뷰모델 객체가 데이터를 업데이트하고 이 업데이트를 상위 모듈에게 노티를 보냈을때 노티를 받은 메소드가 해야할 역할입니다.  반영하였습니다. 피드백 감사합니다.

godrm commented on iOS/Todo/Todo/ViewControllers/MainViewController.swift:
```diff
@@ -29,21 +29,25 @@ final class MainViewController: UIViewController {
         ColumnsUseCase.makeColumns(with: MockColumnsSuccessStub()) { columnsDataSource in
             columnsDataSource?.iterateColumns(with: { column in
                 DispatchQueue.main.async {
```
디스패치 처리도 분리한 하위 함수에 넣으면 안되는건가요?

ehgud0670 replied:
맞습니다. 해당 하위함수의 역할은 뷰를 업데이트하는 메인 스레드가 해야 될 역할이기에 디스패치 처리를 넣어도 됩니다.
따라서 디스패치 처리를 하위함수에 넣어서 반영했는데 반영하다보니 스택에 뷰컨트롤러를 추가하는 메소드`addColumnViewController()`에 뷰 컨트롤러를 생성하는 메소드`columnViewController`도 호출하게 되었습니다. 
처음에는 두 가지 역할을 하는 것 같아 어색했는데, 지금 보니 column 모델을 가지고 뷰 컨트롤러를 스택에 집어넣기 위해 뷰컨트롤러를 생성해달라는 메시지를 요청하고  응답받아  반환된 뷰컨트롤러를 스택에 추가하는거라 합당하다고 봅니다.

ehgud0670 committed:
[#90] refactor: guard else 문에 fatalError 대신 CardCell 객체를 생성하도록
		코드 수정함

		이유: 앱은 에러가 나도 계속 살아있어야 하고 버그를 고치는 것은 그 이후의 일이다.
		      셀 안 만들어진다고 앱이 죽어서는 안된다. 따라서 값이 주입되지 않은 CardCell 객체를 생성해
		      리턴하는 것으로 코드를 수정한 것이다.

ehgud0670 committed:
[#90] [chore] 99자를 넘어가서 개행 정리함

ehgud0670 committed:
[#90] refactor: 유스케이스가 특정한 엔트포인트만을 필요로 하므로 엔드포인트를
		유스케이스가 정적 상수로 갖게 함

ehgud0670 committed:
[#90] refactor: 앤드포인트를 만드는 객체를 생성하여 columnID, cardID를 주면
		urlString을 생성하도록 하였다.

ehgud0670 committed:
[#90] refactor: 데이터소스의 columnID는 삭제하고 뷰컨트롤러가 갖도록 함

		이유: columnID는 뷰에 입히는 데이터가 아니므로 뷰모델로서 데이터 소스가
		안가지고 있어도 되서 columID 프로퍼티를 삭제하였고,
		대신 Column 뷰컨트롤러가 갖는게 맞다고 생각해 Column 뷰컨트롤러가 columnID를 갖도록 했다.

ehgud0670 committed:
[#90] refactor: ColumnDelegate 클래스, 객체를 삭제하고 ColumnViewController가
		UITableViewDelegate의 역할을 하도록 채택하고 이전의 코드를 다 적용함

	이유: 기존 Delegate객체는 이벤트가 발생한 경우 노티를 보내는 역할만 했었고,
		특히 노티에 대한 옵저버가 ColumnViewController 하나밖에 없었다.
		원래 노티피케이션이라는게 1:n관계에서 여러 옵저버에게 보낼때 적합한데
		옵저버가 하나라서 아쉽고(비효율적이다), Delegate 객체도 다른 역할(모델객체를 생성한다거나)은
		하지 않아서 아쉽다(비효율적이다).
		따라서 기존의 델리게이트 객체는 삭제하고, ColumnViewController 객체가 Delegate의 역할을 하여
		좀 더 간단하고 유연한 구조가 되게 만들었다.

ehgud0670 committed:
[#90] refactor: CardViewController 를 파일 분리하고 fileprivate을 private으로 둠
		아직 방법을 못찾아서 internal 접근제한인 것도 있다. 이후 해결할 것이다.

ehgud0670 committed:
[#90] refactor: createCard는 super-class에 있고 subclass 에서 오버라이딩하는게
		적절해서 이렇게 되도록 리팩토링함. 그리고 addTarget은 super class 에서 하도록 함
		* 대신 super-class의 createCard는 비어있다.

ehgud0670 committed:
[#90] refactor: WillEditCardViewModel을 삭제하고 뷰컨트롤러에 따로 row와 CardViewModel을 둬서 사용함
		이유: WillEditCardViewModel을 만든 이유는 순전히 indexPath의 row값이 필요했기 때문이다.
		      row을 알면 이후에 모델 배열에서 해당 row의 모델로 접근하기가 매우 용이하기 때문에 row를 사용한 것인데,
		      row 하나 때문에 또 다른 뷰모델을 만들기 보단 뷰컨트롤러가 따로 row 프로퍼티를 갖는게 더 적절한 것 같아
		      이렇게 코드를 수정하였다.
		      해당 row의 카드를 수정하는 뷰컨트롤러 이므로 해당 row값을 가져도 되지않을까 생각해본다.

ehgud0670 committed:
[#90] refactor: reloadData 메소드는 모델의 값이 바뀐걸 노티를 통해 받았을 때 호출되도록 함
		이유: 이전의 reloadData 메소드가 있던 메소드들은 모델의 값만 바꾸는 역할(엄밀히 말하면 메시지 전달)인것인지
		화면을 갱신하는 역할은 없다. 따라서 이 메소드들에서 화면을 갱신하지않고, 모델 값이 바뀌어서 노티를 받았을 때 화면을 갱신하는 것이 훨씬 합당하고 자연스럽다.
		따라서 코드 변경하였다.

ehgud0670 committed:
[#90] refactor: 디스패 처리도 하위 함수에서 하도록 리팩토링함

ehgud0670 committed:
[#90] refactor: 인코딩을 하위 객체인 NewCard가 하도록 리팩토링함

ehgud0670 committed:
[#90] refactor: 속성들을 캡슐화하고 대신 행동으로 메시지를 주고 받도록 메소드를 구현함

ehgud0670 committed:
[#90] style: 없던 뉴카드 객체를 만드는 것이므로 앞에 generate를 붙임

ehgud0670 commented:
다른 기능을 구현해야돼서 이 풀리퀘는 머지하겠습니다 ㅜ
